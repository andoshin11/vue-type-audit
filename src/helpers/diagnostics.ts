import _ts from 'typescript'
import { isTSVueFile, toRawVueFileName } from './path'
import { pos2location, createSourceFile, location2pos } from './node'
import { SourcemapEntry, DiagnosticWithRange } from '../types'
import { readSystemFile } from './file'
import { findNode } from '../ast'
import { getOriginalPositionFor } from '../sourcemap'

// TS diagnostic code
const REQUIRE_PROPS_CODE = 2554

const EXTERNAL_COMPONENT_IDENTIFIER_REGEXP = /^_component_.*$/

export const hasDiagRange = (diagnostic: _ts.Diagnostic): diagnostic is DiagnosticWithRange => {
  const { start, length, file } = diagnostic
  return typeof start === 'number' && typeof length === 'number' && !!file
}

export const translateVuefileDiagnostic = (diagnostic: _ts.Diagnostic, sourcemapEntry: SourcemapEntry, typeChecker: _ts.TypeChecker, ts: typeof _ts): _ts.Diagnostic => {
  const { file } = diagnostic

  if (!file || !isTSVueFile(file.fileName)) {
    throw new Error('Invalid diagnostic format')
  }

  // Do nothing if the diagnostic has no range
  if (!hasDiagRange(diagnostic)) {
    return diagnostic
  }

  // Update sourcefile
  const rawVueFileName = toRawVueFileName(file.fileName)
  const originalContent = readSystemFile(rawVueFileName)
  if (!originalContent) {
    throw new Error(`could not find file: ${rawVueFileName}`)
  }
  const sourcefile = createSourceFile(ts, originalContent, rawVueFileName)

  // Recover original position
  const { start, length } = diagnostic
  const content = file.getFullText()
  const startLC = pos2location(content, start)
  const endLC = pos2location(content, start + length)

  const originalStartLC = getOriginalPositionFor(file.fileName, startLC, sourcemapEntry)
  const originalEndLC = getOriginalPositionFor(file.fileName, endLC, sourcemapEntry)

  const originalStartPos = location2pos(originalContent, originalStartLC)
  const originalEndPos = location2pos(originalContent, originalEndLC)

  // Translate Message
  const translatedMessage = translateDiagnosticMessage(diagnostic, typeChecker, ts)

  return {
    ...diagnostic,
    file: sourcefile,
    start: originalStartPos,
    length: originalEndPos - originalStartPos,
    messageText: translatedMessage
  }
}

const translateDiagnosticMessage = (diagnostic: DiagnosticWithRange, typeChecker: _ts.TypeChecker, ts: typeof _ts): string | _ts.DiagnosticMessageChain => {
  const { code, start, messageText, file } = diagnostic
  const translationTargetCodes = [REQUIRE_PROPS_CODE]
  let translatedMessage = messageText

  if (!translationTargetCodes.includes(code)) return translatedMessage

  if (diagnostic.code === REQUIRE_PROPS_CODE) {
    const node = findNode(file, start)
    if (!node) return translatedMessage
    const _createVNode = node.parent
    if (!ts.isCallExpression(_createVNode)) return translatedMessage
    const firstArg = _createVNode.arguments[0]
    if (!firstArg) return translatedMessage

    /**
     * Check if it's an external component node
     */

     // Check Node Type
    const typeInfo = typeChecker.getTypeAtLocation(firstArg)
    const isLiteralType = typeInfo.flags === ts.TypeFlags.StringLiteral
    if (!isLiteralType) return translatedMessage

    // Check external component Name
    const isExternalComponentNode = ts.isIdentifier(firstArg) && EXTERNAL_COMPONENT_IDENTIFIER_REGEXP.test(firstArg.escapedText.toString())
    if (!isExternalComponentNode) return translatedMessage

    const tagName = (typeInfo as _ts.LiteralType).value.toString()
    translatedMessage = `Missing props for <${tagName}/> component.`

    // Check external component props type
    const requiredPropsNaemes = findRequiredPropNames(file, tagName, typeChecker, ts)
    if (requiredPropsNaemes.length >= 1) {
      translatedMessage = `Missing props [${requiredPropsNaemes.join(',')}] for <${tagName}/> component.`
    }

    return translatedMessage

  }

  return translatedMessage
}

export function findRequiredPropNames(file: _ts.SourceFile, componentName: string, typeChecker: _ts.TypeChecker, ts: typeof _ts) {
  const findPropsDict = (node: _ts.Node): _ts.Node | undefined => {
    if (ts.isTypeAliasDeclaration(node) && node.name.escapedText.toString() === '___ExternalComponentsProps') {
      return node
    }
    return ts.forEachChild(node, findPropsDict)
  }
  const propsDict = findPropsDict(file)!

  const findTargetProperty = (node: _ts.Node): _ts.Node | undefined => {
    if (ts.isPropertySignature(node) && (node.name as _ts.Identifier).escapedText.toString() === componentName) {
      return node
    }
    return ts.forEachChild(node, findTargetProperty)
  }
  const targetComponent = findTargetProperty(propsDict)!
  const propType = typeChecker.getTypeAtLocation(targetComponent)

  const requiredPropNames = typeChecker.typeToString(propType).split('|').map(s => s.trim())

  return requiredPropNames
}
