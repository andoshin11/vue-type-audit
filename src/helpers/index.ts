export * from './diagnostics'
export * from './node'
export * from './path'
