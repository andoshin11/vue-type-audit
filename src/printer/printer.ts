import _ts from 'typescript'
import * as prettier from 'prettier'

export class Printer {
  printer: _ts.Printer

  constructor(private ts: typeof _ts) {
    this.printer = ts.createPrinter()
  }

  print(nodeList: _ts.Node[]) {
    const { printer, ts } = this
    return printer.printList(
      ts.ListFormat.MultiLine,
      ts.createNodeArray(nodeList),
      ts.createSourceFile('', '', ts.ScriptTarget.ES2015)
    )
  }

  prettyPrint(nodeList: _ts.Node[]) {
    const printed = this.print(nodeList)
    return prettier.format(printed, { parser: 'typescript' })
  }
}
