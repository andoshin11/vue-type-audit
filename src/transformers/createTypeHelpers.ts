function camelize(str: string) {
  return str.charAt(0).toUpperCase() + str.slice(1)
}

export function createTypeHelpers(externalComponents: Record<string, string>, eventNames: string[]) {
  return `
///////////////////// START: Type Helpers ///////////////////
import { ClassInstance, _VNodeProps, _resolveComponent, PropTypes, _VNode, _VNodeTypes, _ClassComponent, _Data, RequiredPropNames, WithEmitType, MixIntoComponent, isNeverType } from "__GLOBAL_TYPES";

type EVENT_DICT = {
${eventNames.map(n => `  on${camelize(n)}: "${n}";`).join('\n')}
}

type __ExternalComponents = {
${Object.entries(externalComponents).map(([key, val]) => `  ${key}: ClassInstance<typeof ${val}>;`).join('\n')}
}

type __ExternalComponentsProps = {
${Object.entries(externalComponents).map(([key, val]) => `  ${key}: RequiredPropNames<__ExternalComponents['${val}']>;`).join('\n')}
}

const ___Component: MixIntoComponent<typeof __Component, WithEmitType<typeof __Component, EVENT_DICT>> = __Component as any

export default ___Component;

type __PropArg<N> = N extends keyof __ExternalComponents ?  PropTypes<__ExternalComponents[N]> : _Data
type __EmitArg<N> = N extends keyof __ExternalComponents ? __ExternalComponents[N]['$options']['__emitHandlerTypes'] : {}

declare function _createVNode<N extends (keyof __ExternalComponents | _VNodeTypes | _ClassComponent)>(type: N, ...args: (N extends keyof __ExternalComponents ? true extends isNeverType<__ExternalComponentsProps[N]> ? [(_VNodeProps & __PropArg<N> & __EmitArg<N>)?, unknown?, number?, (string[] | null)?, boolean?] : [(_VNodeProps & __PropArg<N> & __EmitArg<N>), unknown?, number?, (string[] | null)?, boolean?] : [((_Data & _VNodeProps) | null)?, unknown?, number?, (string[] | null)?, boolean?])): _VNode;
///////////////////// END: Type Helpers /////////////////////
`
}
