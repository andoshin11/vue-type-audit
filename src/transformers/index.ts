export * from './template'
export * from './transformers'
