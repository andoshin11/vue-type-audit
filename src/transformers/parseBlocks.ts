import { parse as parseComponent } from '@vue/compiler-sfc'
import { compile } from '@vue/compiler-dom'
import { RawVueFileName, SourcemapEntry } from '../types'
import { toVueScriptFileName, toVueTemplateFileName, toCompiledVueTemplateFileName } from '../helpers'

export function parseBlocks(
  fileName: RawVueFileName,
  content: string,
  sourcemapEntry: SourcemapEntry
) {
  let { template, script } = parseComponent(content, { sourceMap: true, filename: fileName }).descriptor

  // `template` can be empty, yet you always require `script` block
  if (!script) {
    content += `
<script lang="ts">
import { defineComponent } from 'vue'

export default defineComponent({
})
</script>
    `
    const appended = parseComponent(content, { sourceMap: true, filename: fileName }).descriptor
    template = appended.template
    script = appended.script!
  }

  /**
   * Register sourcemap for blocks
   */

  // `template`
  const vueTemplateFileName = toVueTemplateFileName(fileName)
  // FIXME
  sourcemapEntry.set(vueTemplateFileName, { map: template ? template.map : undefined })

  const compiledVueTemplateFileName = toCompiledVueTemplateFileName(vueTemplateFileName)
  const compiled = compile(template ? template.content : '', { mode: 'module', sourceMap: true, filename: vueTemplateFileName })
  sourcemapEntry.set(compiledVueTemplateFileName, { map: compiled.map })

  // `script`
  const vueScriptFileName = toVueScriptFileName(fileName)
  sourcemapEntry.set(vueScriptFileName, {
    map: {
      ...script.map!,
      file: vueScriptFileName
    }
  })

  let templateContent = !!compiled.code ? compiled.code : undefined
  let scriptContent = script.content

  return {
    templateContent,
    scriptContent
  }
}
