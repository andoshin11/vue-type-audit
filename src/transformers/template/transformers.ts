import _ts from 'typescript'
import { SourceMapGenerator, RawSourceMap } from 'source-map'
import { CompiledVueTemplateFileName } from '../../types'
import { toTransformedCompiledVueTemplateFileName } from '../../helpers'

export function transformTemplate(
  fileName: CompiledVueTemplateFileName,
  content: string
) {
  const transformedCompiledVueTemplateFileName = toTransformedCompiledVueTemplateFileName(fileName)
  const gen = new SourceMapGenerator({ file: transformedCompiledVueTemplateFileName })
  gen.setSourceContent(fileName, content)

  // escape overridden identifiers
  let transformed = content
  transformed = transformed.replace(/_createVNode/, '#createVNode')
  transformed = transformed.replace(/_resolveComponent/, '#resolveComponent')

  // strip export assignment
  transformed = transformed.replace('export function render(_ctx, _cache)', 'function render(_ctx: __ComponentType)')

  transformed
    .split('\n')
    .forEach((_, index) => {
      gen.addMapping({
        source: fileName,
        original: { line: index + 1, column: 0 },
        generated: { line: index + 1, column: 0 }
      })
    })

  const map = JSON.parse(gen.toString()) as RawSourceMap
  return {
    transformed,
    map
  }
}
