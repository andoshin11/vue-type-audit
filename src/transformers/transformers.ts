import _ts from 'typescript'
import * as fs from 'fs'
import { parse as parseComponent } from '@vue/compiler-sfc'
import { compile } from '@vue/compiler-dom'
import { RawVueFileName, SourcemapEntry } from '../types'
import { transformScript } from './script'
import { transformTemplate } from './template'
import { toVueScriptFileName, toTSVueFIleName, toVueTemplateFileName, toCompiledVueTemplateFileName } from '../helpers'

export function transformVueFile(
  fileName: RawVueFileName,
  content: string,
  sourcemapEntry: SourcemapEntry
) {
  const tsVueFileName = toTSVueFIleName(fileName)

  // Parse blocks
  const { template, script } = parseComponent(content, { sourceMap: true, filename: fileName }).descriptor

  // `template` can be empty, yet you always require `script` block
  if (!script) {
    throw new Error('<script/> block is missing in ' + fileName)
  }

  /**
   * Register sourcemap for blocks
   */

  // `template`
  const vueTemplateFileName = toVueTemplateFileName(fileName)
  sourcemapEntry.set(vueTemplateFileName, { map: template ? template.map : undefined })

  const compiledVueTemplateFileName = toCompiledVueTemplateFileName(vueTemplateFileName)
  const compiled = compile(template ? template.content : '', { mode: 'module', sourceMap: true, filename: vueTemplateFileName })
  sourcemapEntry.set(compiledVueTemplateFileName, { map: compiled.map })


  // `script`
  const vueScriptFileName = toVueScriptFileName(fileName)
  sourcemapEntry.set(vueScriptFileName, {
    map: {
      ...script.map!,
      file: vueScriptFileName
    }
  })

  let templateContent = !!compiled.code ? compiled.code : undefined
  let scriptContent = script.content


  /**
   * Create outputs
   */
  const preparationBlock = createPreparationBlock()

  const scriptBlock = transformScript(vueScriptFileName, scriptContent, _ts)
  const templateBlock = transformTemplate(compiledVueTemplateFileName, templateContent)

  const preparationBlockLinesNum = preparationBlock.split('\n').length
  const scriptBlockLinesNum = preparationBlockLinesNum + 1 + scriptBlock.transformed.split('\n').length

  const map = {
    version: '3',
    file: tsVueFileName,
    sections: [
      // script block
      {
        offset: {
          line: preparationBlockLinesNum,
          column: 0
        },
        map: scriptBlock.map
      },
      // template block
      {
        offset: {
          line: scriptBlockLinesNum,
          column: 0
        },
        map: templateBlock.map
      }
    ]
  } as any

  sourcemapEntry.set(tsVueFileName, { map })

  const output = preparationBlock + '\n' + scriptBlock.transformed + '\n' + templateBlock.transformed
  fs.writeFileSync(toTSVueFIleName(fileName), output)

  return output
}

function createPreparationBlock() {
  return 'import { ClassInstance, __createVNodeFromExternalComponent, _VNodeProps, PropTypes, _VNode, _VNodeTypes, _ClassComponent, _Data } from "__GLOBAL_TYPES";'
}
