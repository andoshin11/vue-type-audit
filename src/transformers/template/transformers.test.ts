import ts from 'typescript'
import { createSourceFile } from '../../helpers'
import { Printer } from '../../printer'
import { transformRenderCall } from './transformers'

describe('transformers/template/transformers', () => {
  const printer = new Printer(ts)


  test('transformTemplate', () => {
    const rawText = `
import { toDisplayString as _toDisplayString, createVNode as _createVNode, openBlock as _openBlock, createBlock as _createBlock } from "vue";

export function render(_ctx, _cache) {
  return (_openBlock(), _createBlock("div", null, [
    _createVNode("h1", null, _toDisplayString(_ctx.msg), 1 /* TEXT */)
  ]))
}
    `
    const sourceFile = createSourceFile(ts, rawText, ts.ScriptTarget.ES2017)
    const output = transformRenderCall(sourceFile, ts)

    expect(printer.print(output)).toMatchSnapshot()
  })

  test('transformRenderCall', () => {
    const rawText = `
import { toDisplayString as _toDisplayString, createVNode as _createVNode, openBlock as _openBlock, createBlock as _createBlock } from "vue";

export function render(_ctx, _cache) {
  return (_openBlock(), _createBlock("div", null, [
    _createVNode("h1", null, _toDisplayString(_ctx.msg), 1 /* TEXT */)
  ]))
}
    `
    const sourceFile = createSourceFile(ts, rawText, ts.ScriptTarget.ES2017)
    const output = transformRenderCall(sourceFile, ts)

    expect(printer.print(output)).toMatchSnapshot()
  })
})