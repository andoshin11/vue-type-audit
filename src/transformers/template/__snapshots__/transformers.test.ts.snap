// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transformers/template/transformers transformTemplate 1`] = `
Object {
  "map": Object {
    "file": "test.vue.template.ts",
    "mappings": "AAAA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA",
    "names": Array [],
    "sources": Array [
      "test.vue.template.ts",
    ],
    "sourcesContent": Array [
      "
import { toDisplayString as _toDisplayString, createVNode as _createVNode, openBlock as _openBlock, createBlock as _createBlock } from \\"vue\\";

export function render(_ctx, _cache) {
  return (_openBlock(), _createBlock(\\"div\\", null, [
    _createVNode(\\"h1\\", null, _toDisplayString(_ctx.msg), 1 /* TEXT */)
  ]))
}
    ",
    ],
    "version": 3,
  },
  "transformed": "
import { toDisplayString as _toDisplayString, createVNode as #createVNode, openBlock as _openBlock, createBlock as _createBlock } from \\"vue\\";

function render(_ctx: ClassInstance<typeof __Component>) {
  return (_openBlock(), _createBlock(\\"div\\", null, [
    _createVNode(\\"h1\\", null, _toDisplayString(_ctx.msg), 1 /* TEXT */)
  ]))
}
    ",
}
`;

exports[`transformers/template/transformers transformTemplate 2`] = `
Object {
  "map": Object {
    "file": "test.vue.template.ts",
    "mappings": "AAAA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA",
    "names": Array [],
    "sources": Array [
      "test.vue.template.ts",
    ],
    "sourcesContent": Array [
      "
import { createVNode as _createVNode, toDisplayString as _toDisplayString, resolveComponent as _resolveComponent, openBlock as _openBlock, createBlock as _createBlock } from \\"vue\\"

function render(_ctx: __ComponentType) {
  const _component_HelloWorld = _resolveComponent(\\"HelloWorld\\")

  return (_openBlock(), _createBlock(\\"div\\", null, [
    _createVNode(\\"img\\", { src: \\"./logo.png\\" }),
    _createVNode(\\"h1\\", null, \\"Hello Vue 3!\\"),
    _createVNode(\\"button\\", { onClick: _ctx.inc }, \\"Clicked \\" + _toDisplayString(_ctx.count.length) + \\" times.\\", 9 /* TEXT, PROPS */, [\\"onClick\\"]),
    _createVNode(_component_HelloWorld)
  ]))
}   
    ",
    ],
    "version": 3,
  },
  "transformed": "
import { createVNode as #createVNode, toDisplayString as _toDisplayString, resolveComponent as #resolveComponent, openBlock as _openBlock, createBlock as _createBlock } from \\"vue\\"

function render(_ctx: __ComponentType) {
  const _component_HelloWorld = _resolveComponent(\\"HelloWorld\\")

  return (_openBlock(), _createBlock(\\"div\\", null, [
    _createVNode(\\"img\\", { src: \\"./logo.png\\" }),
    _createVNode(\\"h1\\", null, \\"Hello Vue 3!\\"),
    _createVNode(\\"button\\", { onClick: _ctx.inc }, \\"Clicked \\" + _toDisplayString(_ctx.count.length) + \\" times.\\", 9 /* TEXT, PROPS */, [\\"onClick\\"]),
    _createVNode(_component_HelloWorld)
  ]))
}   
    ",
}
`;
